src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    /// <summary>
+    /// Middleware that logs HTTP requests and HTTP responses.
+    /// </summary>
+    internal sealed class W3CLoggingMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly W3CLogger _w3cLogger;
+        private readonly IOptionsMonitor<W3CLoggerOptions> _options;
+
+        /// <summary>
+        /// Initializes <see cref="W3CLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="options"></param>
+        /// <param name="environment"></param>
+        public W3CLoggingMiddleware(RequestDelegate next, IOptionsMonitor<W3CLoggerOptions> options, IHostEnvironment environment)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (environment == null)
+            {
+                throw new ArgumentNullException(nameof(environment));
+            }
+
+            _options = options;
+            if (String.IsNullOrEmpty(_options.CurrentValue.LogDirectory))
+            {
+                _options.CurrentValue.LogDirectory = Path.Join(environment.ContentRootPath, "logs");
+            }
+            _w3cLogger = new W3CLogger(_options);
+        }
+
+        /// <summary>
+        /// Invokes the <see cref="HttpLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>HttpResponseLog.cs
+        public async Task Invoke(HttpContext context)
+        {
+            var options = _options.CurrentValue;
+
+            var w3cList = new List<KeyValuePair<string, object?>>();
```

Comment:
I think in this middleware there's nothing against using `List<KeyValuePair<string, string?>>`.
```suggestion
            var w3cList = new List<KeyValuePair<string, string?>>();
```

Reply:
My thinking was to keep it as `string, object` in case we ever need to update `W3CLogger` to implement `ILogger` - but it might be better cross that bridge if/when we come to it (meaning it might be better to switch to `string, string` now). I'll push that once I clean up a couple other things